zingo-sync/src/scan/workers.rs

```diff
@@ -0,0 +1,173 @@
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{self, AtomicBool},
+        Arc,
+    },
+};
+
+use tokio::{sync::mpsc, task::JoinHandle};
+
+use zcash_client_backend::data_api::scanning::ScanRange;
+use zcash_keys::keys::UnifiedFullViewingKey;
+use zcash_primitives::{consensus::Parameters, zip32::AccountId};
+
+use crate::{client::FetchRequest, keys::ScanningKeys, primitives::WalletBlock};
+
+use super::{scan, ScanResults};
+
+const SCAN_WORKER_POOLSIZE: usize = 2;
+
+pub(crate) struct Scanner<P> {
+    workers: Vec<WorkerHandle>,
+    scan_results_sender: mpsc::UnboundedSender<(ScanRange, ScanResults)>,
+    fetch_request_sender: mpsc::UnboundedSender<FetchRequest>,
+    parameters: P,
+    ufvks: HashMap<AccountId, UnifiedFullViewingKey>,
+}
+
+// TODO: add fn for checking and handling worker errors
+impl<P> Scanner<P>
+where
+    P: Parameters + Sync + Send + 'static,
+{
+    pub(crate) fn new(
+        scan_results_sender: mpsc::UnboundedSender<(ScanRange, ScanResults)>,
+        fetch_request_sender: mpsc::UnboundedSender<FetchRequest>,
+        parameters: P,
+        ufvks: HashMap<AccountId, UnifiedFullViewingKey>,
+    ) -> Self {
+        let workers: Vec<WorkerHandle> = Vec::with_capacity(SCAN_WORKER_POOLSIZE);
+
+        Self {
+            workers,
+            scan_results_sender,
+            fetch_request_sender,
+            parameters,
+            ufvks,
+        }
+    }
+
+    pub(crate) fn spawn_workers(&mut self) {
+        for _ in 0..SCAN_WORKER_POOLSIZE {
+            let (scan_task_sender, scan_task_receiver) = mpsc::unbounded_channel();
+            let worker = ScanWorker::new(
+                scan_task_receiver,
+                self.scan_results_sender.clone(),
+                self.fetch_request_sender.clone(),
+                self.parameters.clone(),
+                self.ufvks.clone(),
+            );
+            let is_scanning = Arc::clone(&worker.is_scanning);
+            let handle = tokio::spawn(async move { worker.run().await });
+            self.workers.push(WorkerHandle {
+                _handle: handle,
+                is_scanning,
+                scan_task_sender,
+            });
+        }
+    }
+
+    pub(crate) fn is_worker_idle(&self) -> bool {
+        self.workers.iter().any(|worker| !worker.is_scanning())
+    }
+
+    pub(crate) fn add_scan_task(&self, scan_task: ScanTask) -> Result<(), ()> {
+        if let Some(worker) = self.workers.iter().find(|worker| !worker.is_scanning()) {
+            worker.add_scan_task(scan_task);
+        } else {
+            panic!("no idle workers!")
```

Comment:
to be replaced with err handling

Reply:
yes! all panics and asserts and unwraps will definately be replaced after PoC